File: apps/desktop/src/deploy/args.ts
export interface SshTarget {
  host: string;
  sshPort?: number;
  sshPassword?: string;
}
export type DeployMethod = "native" | "docker";
export interface DeployRequest {
  target: SshTarget;
  method: DeployMethod;
  version: string;
  listen: string;
  bundleUrl?: string;
}
export interface DeployBrand {
  envPrefix: string;
  daemonPort: number;
  releaseBase: string | null;
}

export function record(input: unknown): Record<string, unknown> {
  if (!input || typeof input !== "object" || Array.isArray(input))
    throw new Error("Deploy arguments must be an object.");
  return input as Record<string, unknown>;
}
function text(input: unknown): string {
  return typeof input === "string" ? input.trim() : "";
}
function hasUnsafeCharacters(value: string, forbidden: RegExp): boolean {
  return (
    forbidden.test(value) ||
    [...value].some((character) => {
      const code = character.charCodeAt(0);
      return code < 32 || code === 127;
    })
  );
}
export function parseTarget(input: unknown): SshTarget {
  const args = record(input);
  const host = text(args.host);
  if (!host) throw new Error("SSH host is required");
  if (host.startsWith("-") || hasUnsafeCharacters(host, /\s/u))
    throw new Error("SSH host is invalid");
  const sshPort = args.sshPort;
  if (
    sshPort != null &&
    (typeof sshPort !== "number" || !Number.isInteger(sshPort) || sshPort < 1 || sshPort > 65535)
  )
    throw new Error("SSH port must be between 1 and 65535.");
  return {
    host,
    ...(sshPort == null ? {} : { sshPort: sshPort as number }),
    ...(typeof args.sshPassword === "string" && args.sshPassword
      ? { sshPassword: args.sshPassword }
      : {}),
  };
}
export function parseMethod(input: unknown): DeployMethod {
  const method = text(record(input).method) || "native";
  if (method !== "native" && method !== "docker")
    throw new Error(`Unknown deploy method: ${method}`);
  return method;
}
export function parseRequest(
  input: unknown,
  defaultVersion: string,
  brand: DeployBrand,
): DeployRequest {
  const args = record(input);
  const target = parseTarget(args);
  const method = parseMethod(args);
  const version = (text(args.version) || defaultVersion).replace(/^v+/u, "");
  if (!/^[A-Za-z0-9.+_-]{1,64}$/u.test(version)) throw new Error("Version is invalid");
  const listen = text(args.listen) || `0.0.0.0:${brand.daemonPort}`;
  const split = listen.lastIndexOf(":");
  const host = listen.slice(0, split);
  const port = listen.slice(split + 1);
  if (
    split < 1 ||
    hasUnsafeCharacters(host, /[\s'"]/u) ||
    !/^\d+$/u.test(port) ||
    Number(port) < 1 ||
    Number(port) > 65535
  )
    throw new Error("Listen address must be host:port");
  const bundleUrl = text(args.bundleUrl) || undefined;
  if (bundleUrl) {
    let url: URL;
    try {
      url = new URL(bundleUrl);
    } catch {
      throw new Error("Bundle URL must be an http(s) URL");
    }
    if (
      !["http:", "https:"].includes(url.protocol) ||
      url.username ||
      url.password ||
      bundleUrl.length > 2048 ||
      hasUnsafeCharacters(bundleUrl, /[\s']/u)
    )
      throw new Error("Bundle URL must be an http(s) URL");
  }
  return { target, method, version, listen, bundleUrl };
}
export function shellQuote(value: string): string {
  return `'${value.replaceAll("'", "'\\''")}'`;
}
export function buildInstallCommand(request: DeployRequest, brand: DeployBrand): string {
  const values: Record<string, string> = { VERSION: request.version };
  if (request.method === "native") {
    values.LISTEN = request.listen;
    if (brand.releaseBase) values.RELEASE_BASE = brand.releaseBase;
    if (request.bundleUrl) values.BUNDLE_URL = request.bundleUrl;
    if (!brand.releaseBase && !request.bundleUrl)
      throw new Error("This distribution has no daemon release URL.");
  } else {
    const split = request.listen.lastIndexOf(":");
    values.BIND = request.listen.slice(0, split);
    values.PORT = request.listen.slice(split + 1);
  }
  return `${Object.entries(values)
    .map(([key, value]) => `${brand.envPrefix}_${key}=${shellQuote(value)}`)
    .join(" ")} bash -s`;
}
export function buildSshArgs(target: SshTarget, command: string): string[] {
  return [
    "-T",
    "-o",
    ...(target.sshPassword
      ? [
          "NumberOfPasswordPrompts=1",
          "-o",
          "PreferredAuthentications=publickey,keyboard-interactive,password",
        ]
      : ["BatchMode=yes"]),
    "-o",
    "ConnectTimeout=10",
    ...(target.sshPort ? ["-p", String(target.sshPort)] : []),
    target.host,
    command,
  ];
}

File: apps/desktop-tauri/src-tauri/src/deploy/args.rs
//! Argument parsing for the `ssh_deploy_*` commands and the shell-side
//! encoding of their values. Host strings follow the transport's rules (no
//! leading `-`, no whitespace) so a host can never become an ssh option; every
//! value that reaches the remote command line is single-quoted.

use serde_json::Value;

use crate::transport::ssh_auth::{password_from_args, SshPassword};

pub const DEFAULT_LISTEN: &str = crate::branding::DEFAULT_LISTEN;
pub const DEFAULT_RELEASE_BASE: &str = crate::branding::RELEASE_BASE;
const MAX_VERSION_LEN: usize = 64;
const MAX_URL_LEN: usize = 2048;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DeployMethod {
    Native,
    Docker,
}

impl DeployMethod {
    pub fn as_str(self) -> &'static str {
        match self {
            DeployMethod::Native => "native",
            DeployMethod::Docker => "docker",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SshTarget {
    pub host: String,
    pub ssh_port: Option<u16>,
    /// Answers ssh's password prompt through the askpass helper
    /// (`transport::ssh_auth`); absent means `BatchMode`.
    pub ssh_password: Option<SshPassword>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DeployRequest {
    pub target: SshTarget,
    pub method: DeployMethod,
    pub version: String,
    pub listen: String,
    pub bundle_url: Option<String>,
}

/// Same rules as `validate_ssh_host` in `transport::session`.
pub fn validate_ssh_host(raw: &str) -> Result<String, String> {
    let host = raw.trim();
    if host.is_empty() {
        return Err("SSH host is required".into());
    }
    if host.chars().any(char::is_whitespace) || host.starts_with('-') {
        return Err("SSH host is invalid".into());
    }
    Ok(host.to_string())
}

fn parse_port(value: Option<&Value>, label: &str) -> Result<Option<u16>, String> {
    match value {
        None | Some(Value::Null) => Ok(None),
        Some(Value::Number(number)) => number
            .as_u64()
            .filter(|port| (1..=65535).contains(port))
            .map(|port| Some(port as u16))
            .ok_or_else(|| format!("{label} must be between 1 and 65535.")),
        Some(_) => Err(format!("{label} must be between 1 and 65535.")),
    }
}

fn string_field<'a>(args: &'a Value, key: &str) -> Option<&'a str> {
    args.get(key)
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
}

pub fn parse_ssh_target(args: &Value) -> Result<SshTarget, String> {
    if !args.is_object() {
        return Err("Deploy arguments must be an object.".into());
    }
    Ok(SshTarget {
        host: validate_ssh_host(string_field(args, "host").unwrap_or_default())?,
        ssh_port: parse_port(args.get("sshPort"), "SSH port")?,
        ssh_password: password_from_args(args),
    })
}

pub fn parse_method(args: &Value) -> Result<DeployMethod, String> {
    match string_field(args, "method") {
        None | Some("native") => Ok(DeployMethod::Native),
        Some("docker") => Ok(DeployMethod::Docker),
        Some(other) => Err(format!("Unknown deploy method: {other}")),
    }
}

/// A release version or image tag: the characters a semver string or a
/// Docker tag can contain, nothing that means anything to a shell.
pub fn validate_version(raw: &str) -> Result<String, String> {
    let version = raw.trim().trim_start_matches('v');
    let valid = !version.is_empty()
        && version.len() <= MAX_VERSION_LEN
        && version
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, '.' | '-' | '+' | '_'));
    if valid {
        Ok(version.to_string())
    } else {
        Err("Version is invalid".into())
    }
}

/// `host:port` with a numeric port; the host part may not be empty.
pub fn validate_listen(raw: &str) -> Result<String, String> {
    let listen = raw.trim();
    let (host, port) = split_listen(listen).ok_or("Listen address must be host:port")?;
    if host.is_empty()
        || host
            .chars()
            .any(|c| c.is_whitespace() || c == '\'' || c == '"')
        || !(1..=65535).contains(&port)
    {
        return Err("Listen address must be host:port".into());
    }
    Ok(listen.to_string())
}

/// Splits `host:port` at the last colon (so `[::1]:9999` keeps its host).
pub fn split_listen(listen: &str) -> Option<(&str, u16)> {
    let (host, port) = listen.rsplit_once(':')?;
    let port: u16 = port.parse().ok()?;
    Some((host, port))
}

pub fn validate_bundle_url(raw: &str) -> Result<String, String> {
    let url = raw.trim();
    let valid = (url.starts_with("https://") || url.starts_with("http://"))
        && url.len() <= MAX_URL_LEN
        && !url
            .chars()
            .any(|c| c.is_whitespace() || c.is_control() || c == '\'');
    if valid {
        Ok(url.to_string())
    } else {
        Err("Bundle URL must be an http(s) URL".into())
    }
}

pub fn parse_deploy_request(args: &Value, default_version: &str) -> Result<DeployRequest, String> {
    let target = parse_ssh_target(args)?;
    let method = parse_method(args)?;
    let version = validate_version(string_field(args, "version").unwrap_or(default_version))?;
    let listen = validate_listen(string_field(args, "listen").unwrap_or(DEFAULT_LISTEN))?;
    let bundle_url = string_field(args, "bundleUrl")
        .map(validate_bundle_url)
        .transpose()?;
    Ok(DeployRequest {
        target,
        method,
        version,
        listen,
        bundle_url,
    })
}

/// POSIX single-quoting: safe for any byte sequence, including newlines.
pub fn shell_quote(value: &str) -> String {
    format!("'{}'", value.replace('\'', "'\\''"))
}

/// `KEY='value' KEY2='value2'` prefix for a remote command line.
pub fn env_assignments(pairs: &[(&str, &str)]) -> String {
    pairs
        .iter()
        .map(|(key, value)| {
            format!(
                "{}={}",
                key.strip_prefix("FROGG_")
                    .map(crate::branding::env_key)
                    .unwrap_or_else(|| key.to_string()),
                shell_quote(value)
            )
        })
        .collect::<Vec<_>>()
        .join(" ")
}

/// The remote command that reads the install script from stdin.
pub fn build_install_command(request: &DeployRequest) -> String {
    let mut pairs: Vec<(&str, &str)> = vec![("FROGG_VERSION", request.version.as_str())];
    let (bind, port) =
        split_listen(&request.listen).unwrap_or(("127.0.0.1", crate::branding::DEFAULT_PORT));
    let port = port.to_string();
    match request.method {
        DeployMethod::Native => {
            pairs.push(("FROGG_LISTEN", request.listen.as_str()));
            pairs.push(("FROGG_RELEASE_BASE", DEFAULT_RELEASE_BASE));
            if let Some(url) = &request.bundle_url {
                pairs.push(("FROGG_BUNDLE_URL", url.as_str()));
            }
        }
        DeployMethod::Docker => {
            pairs.push(("FROGG_BIND", bind));
            pairs.push(("FROGG_PORT", port.as_str()));
        }
    }
    format!("{} bash -s", env_assignments(&pairs))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn request(method: DeployMethod) -> DeployRequest {
        DeployRequest {
            target: SshTarget {
                host: "dev@box".into(),
                ssh_port: None,
                ssh_password: None,
            },
            method,
            version: "0.2.0".into(),
            listen: "127.0.0.1:9999".into(),
            bundle_url: None,
        }
    }

    #[test]
    fn host_rules_match_transport() {
        assert_eq!(validate_ssh_host(" box ").unwrap(), "box");
        assert!(validate_ssh_host("").is_err());
        assert!(validate_ssh_host("-oProxyCommand=x").is_err());
        assert!(validate_ssh_host("a b").is_err());
    }

    #[test]
    fn quotes_values_for_posix_sh() {
        assert_eq!(shell_quote("plain"), "'plain'");
        assert_eq!(shell_quote("it's"), "'it'\\''s'");
        assert_eq!(shell_quote("$(rm -rf /)"), "'$(rm -rf /)'");
        assert_eq!(
            env_assignments(&[("A", "1"), ("B", "x y")]),
            "A='1' B='x y'"
        );
    }

    #[test]
    fn builds_native_and_docker_commands() {
        assert_eq!(
            build_install_command(&request(DeployMethod::Native)),
            "FROGG_VERSION='0.2.0' FROGG_LISTEN='127.0.0.1:9999' \
             FROGG_RELEASE_BASE='https://github.com/frogg-app/frogg/releases' bash -s"
        );
        let mut with_url = request(DeployMethod::Native);
        with_url.bundle_url = Some("https://example.com/b.tar.gz".into());
        assert!(build_install_command(&with_url)
            .contains("FROGG_BUNDLE_URL='https://example.com/b.tar.gz' bash -s"));
        let mut docker = request(DeployMethod::Docker);
        docker.listen = "0.0.0.0:7000".into();
        assert_eq!(
            build_install_command(&docker),
            "FROGG_VERSION='0.2.0' FROGG_BIND='0.0.0.0' FROGG_PORT='7000' bash -s"
        );
    }

    #[test]
    fn parses_request_with_defaults_and_rejects_bad_values() {
        let parsed = parse_deploy_request(&json!({ "host": "box" }), "0.1.6").unwrap();
        assert_eq!(parsed.method, DeployMethod::Native);
        assert_eq!(parsed.version, "0.1.6");
        assert_eq!(parsed.listen, DEFAULT_LISTEN);
        assert_eq!(parsed.bundle_url, None);

        let parsed = parse_deploy_request(
            &json!({ "host": "box", "sshPort": 2222, "method": "docker", "version": "v1.2.3",
                     "listen": "0.0.0.0:9999", "bundleUrl": "https://x.y/z.tar.gz" }),
            "0.1.6",
        )
        .unwrap();
        assert_eq!(parsed.target.ssh_port, Some(2222));
        assert_eq!(parsed.target.ssh_password, None);
        assert_eq!(parsed.method, DeployMethod::Docker);

        let with_password =
            parse_ssh_target(&json!({ "host": "box", "sshPassword": "hunter2" })).unwrap();
        assert_eq!(
            with_password.ssh_password.as_ref().map(SshPassword::expose),
            Some("hunter2")
        );
        assert!(!format!("{with_password:?}").contains("hunter2"));
        assert_eq!(parsed.version, "1.2.3");
        assert_eq!(parsed.bundle_url.as_deref(), Some("https://x.y/z.tar.gz"));

        assert!(parse_deploy_request(&json!({ "host": "box", "method": "snap" }), "0").is_err());
        assert!(parse_deploy_request(&json!({ "host": "box", "version": "1;rm" }), "0").is_err());
        assert!(parse_deploy_request(&json!({ "host": "box", "listen": "nope" }), "0").is_err());
        assert!(parse_deploy_request(&json!({ "host": "box", "listen": ":0" }), "0").is_err());
        assert!(
            parse_deploy_request(&json!({ "host": "box", "bundleUrl": "ftp://x" }), "0").is_err()
        );
        assert!(parse_deploy_request(&json!({ "host": "box", "sshPort": 70000 }), "0").is_err());
        assert!(parse_deploy_request(&json!("box"), "0").is_err());
    }

    #[test]
    fn listen_split_keeps_ipv6_host() {
        assert_eq!(split_listen("[::1]:9999"), Some(("[::1]", 9999)));
        assert_eq!(split_listen("9999"), None);
        assert!(validate_listen("[::1]:9999").is_ok());
    }
}
